File: src/services/sns.js
// services/sns.js
import { store, randomId, arn } from '../store.js';
import { xmlResponse, errorXml, escapeXml, getRawBody } from '../middleware/response.js';

const ACCOUNT = '000000000000';

export async function handler(req, res) {
  const body = getRawBody(req);
  const params = new URLSearchParams(body);
  const action = req.url.includes('?') ? new URL(req.url,'http://x').searchParams.get('Action') : params.get('Action');

  switch (action) {
    case 'CreateTopic': {
      const name = params.get('Name');
      const a = arn('sns', name);
      if (!store.sns.topics[a]) store.sns.topics[a] = { name, arn: a, created: Date.now(), published: 0, subscriptions: [], attributes: {} };
      return xmlResponse(res, 200, wrap('CreateTopicResponse','CreateTopicResult', `<TopicArn>${escapeXml(a)}</TopicArn>`));
    }
    case 'DeleteTopic': {
      const a = params.get('TopicArn');
      delete store.sns.topics[a];
      return xmlResponse(res, 200, wrap('DeleteTopicResponse','DeleteTopicResult',''));
    }
    case 'ListTopics': {
      const topics = Object.values(store.sns.topics).map(t => `<member><TopicArn>${escapeXml(t.arn)}</TopicArn></member>`).join('');
      return xmlResponse(res, 200, wrap('ListTopicsResponse','ListTopicsResult', `<Topics>${topics}</Topics>`));
    }
    case 'GetTopicAttributes': {
      const topic = store.sns.topics[params.get('TopicArn')];
      if (!topic) return errorXml(res, 404, 'NotFound', 'Topic not found');
      const attrs = {
        TopicArn: topic.arn,
        Owner: ACCOUNT,
        SubscriptionsConfirmed: String(topic.subscriptions.length),
        SubscriptionsPending: '0',
        SubscriptionsDeleted: '0',
        DisplayName: topic.name,
        ...topic.attributes,
      };
      return xmlResponse(res, 200, wrap('GetTopicAttributesResponse','GetTopicAttributesResult', `<Attributes>${attrEntriesXml(attrs)}</Attributes>`));
    }
    case 'SetTopicAttributes': {
      const topic = store.sns.topics[params.get('TopicArn')];
      if (!topic) return errorXml(res, 404, 'NotFound', 'Topic not found');
      const name = params.get('AttributeName');
      if (name) { (topic.attributes ??= {})[name] = params.get('AttributeValue') ?? ''; }
      return xmlResponse(res, 200, wrap('SetTopicAttributesResponse','SetTopicAttributesResult',''));
    }
    case 'Subscribe': {
      const topicArn = params.get('TopicArn');
      const protocol = params.get('Protocol');
      const endpoint = params.get('Endpoint');
      const topic = store.sns.topics[topicArn];
      if (!topic) return errorXml(res, 404, 'NotFound', 'Topic not found');
      const subArn = `${topicArn}:${randomId(8)}`;
      const sub = { subArn, topicArn, protocol, endpoint, status: 'confirmed', owner: ACCOUNT };
      applySubscriptionAttributes(sub, parseStringMap(params, 'Attributes'));
      topic.subscriptions.push(sub);
      return xmlResponse(res, 200, wrap('SubscribeResponse','SubscribeResult', `<SubscriptionArn>${escapeXml(subArn)}</SubscriptionArn>`));
    }
    case 'SetSubscriptionAttributes': {
      const subArn = params.get('SubscriptionArn');
      const sub = findSubscription(subArn);
      if (!sub) return errorXml(res, 404, 'NotFound', 'Subscription not found');
      const name = params.get('AttributeName');
      if (name) applySubscriptionAttributes(sub, { [name]: params.get('AttributeValue') ?? '' });
      return xmlResponse(res, 200, wrap('SetSubscriptionAttributesResponse','SetSubscriptionAttributesResult',''));
    }
    case 'GetSubscriptionAttributes': {
      const sub = findSubscription(params.get('SubscriptionArn'));
      if (!sub) return errorXml(res, 404, 'NotFound', 'Subscription not found');
      const attrs = {
        SubscriptionArn: sub.subArn,
        TopicArn: sub.topicArn,
        Protocol: sub.protocol,
        Endpoint: sub.endpoint,
        Owner: sub.owner || ACCOUNT,
        ConfirmationWasAuthenticated: 'true',
        PendingConfirmation: 'false',
        RawMessageDelivery: sub.rawMessageDelivery ? 'true' : 'false',
        ...(sub.filterPolicy ? { FilterPolicy: sub.filterPolicy, FilterPolicyScope: sub.filterPolicyScope || 'MessageAttributes' } : {}),
      };
      return xmlResponse(res, 200, wrap('GetSubscriptionAttributesResponse','GetSubscriptionAttributesResult', `<Attributes>${attrEntriesXml(attrs)}</Attributes>`));
    }
    case 'Publish': {
      const topicArn = params.get('TopicArn');
      const topic = store.sns.topics[topicArn];
      if (!topic) return errorXml(res, 404, 'NotFound', 'Topic not found');
      const message = params.get('Message') || '';
      const subject = params.get('Subject') || '';
      const attributes = parseMessageAttributes(params, 'MessageAttributes');
      topic.published++;
      const msgId = randomId(36);
      // Fan out to subscribers — async, not awaited (matches AWS semantics:
      // Publish returns success once the message is durably accepted, not
      // once delivered).
      fanoutSnsMessage(topic, { msgId, message, subject, attributes }).catch(e => {
        console.warn(`[SNS] Fanout error for ${topicArn}:`, e.message);
      });
      return xmlResponse(res, 200, wrap('PublishResponse','PublishResult', `<MessageId>${msgId}</MessageId>`));
    }
    case 'PublishBatch': {
      const topicArn = params.get('TopicArn');
      const topic = store.sns.topics[topicArn];
      if (!topic) return errorXml(res, 404, 'NotFound', 'Topic not found');
      const entries = parsePublishBatchEntries(params);
      const successful = [];
      for (const e of entries) {
        topic.published++;
        const msgId = randomId(36);
        fanoutSnsMessage(topic, { msgId, message: e.Message, subject: e.Subject, attributes: e.MessageAttributes })
          .catch(err => console.warn(`[SNS] Fanout error for ${topicArn}:`, err.message));
        successful.push({ Id: e.Id, MessageId: msgId });
      }
      const successXml = successful.map(s => `<member><Id>${escapeXml(s.Id)}</Id><MessageId>${s.MessageId}</MessageId></member>`).join('');
      return xmlResponse(res, 200, wrap('PublishBatchResponse','PublishBatchResult',
        `<Successful>${successXml}</Successful><Failed/>`));
    }
    case 'Unsubscribe': {
      const subArn = params.get('SubscriptionArn');
      // SubscriptionArn format we generate: `${topicArn}:${randomId(8)}`
      // — find the parent topic and remove the matching subscription.
      for (const topic of Object.values(store.sns.topics)) {
        topic.subscriptions = (topic.subscriptions || []).filter(s => s.subArn !== subArn);
      }
      return xmlResponse(res, 200, wrap('UnsubscribeResponse','UnsubscribeResult',''));
    }
    case 'ListSubscriptions': {
      const all = Object.values(store.sns.topics).flatMap(t => t.subscriptions.map(s => ({...s, topicArn: t.arn})));
      return xmlResponse(res, 200, wrap('ListSubscriptionsResponse','ListSubscriptionsResult',
        `<Subscriptions>${all.map(subMemberXml).join('')}</Subscriptions>`));
    }
    case 'ListSubscriptionsByTopic': {
      const topic = store.sns.topics[params.get('TopicArn')];
      if (!topic) return errorXml(res, 404, 'NotFound', 'Topic not found');
      const subs = topic.subscriptions.map(s => ({ ...s, topicArn: topic.arn }));
      return xmlResponse(res, 200, wrap('ListSubscriptionsByTopicResponse','ListSubscriptionsByTopicResult',
        `<Subscriptions>${subs.map(subMemberXml).join('')}</Subscriptions>`));
    }
    default:
      return errorXml(res, 400, 'InvalidAction', `Unknown action: ${action}`);
  }
}

// Fan out a published message to all topic subscribers. Each subscription
// is delivered independently — failure of one doesn't block others.
// Exported so EventBridge SNS targets + S3 notifications can re-trigger fanout.
//   attributes: internal message-attribute map { name: { DataType, StringValue|BinaryValue } }
export async function fanoutSnsMessage(topic, { msgId, message, subject, attributes }) {
  const subs = topic.subscriptions || [];
  const snsEnvelope = {
    Type:      'Notification',
    MessageId: msgId,
    TopicArn:  topic.arn,
    Subject:   subject,
    Message:   message,
    Timestamp: new Date().toISOString(),
    ...(hasAttributes(attributes) ? { MessageAttributes: toEnvelopeAttributes(attributes) } : {}),
  };

  // Lazy imports to avoid circular deps at module load
  const [{ invokeLambda }, { enqueueMessage, queueUrlForArn }] = await Promise.all([
    import('./lambda.js'),
    import('./sqs.js'),
  ]);

  for (const sub of subs) {
    try {
      // Subscription filter policy — skip subscribers whose policy doesn't match.
      if (!subscriptionMatches(sub, attributes, message)) continue;

      // SQS subscription — endpoint is the queue ARN
      if (sub.protocol === 'sqs' || sub.endpoint?.includes(':sqs:')) {
        const url = queueUrlForArn(sub.endpoint);
        if (url && store.sqs.queues[url]) {
          // Raw message delivery: enqueue the bare Message (carrying the
          // original attributes) instead of the JSON SNS envelope.
          if (sub.rawMessageDelivery) enqueueMessage(url, message, { attributes });
          else enqueueMessage(url, JSON.stringify(snsEnvelope));
        }
        continue;
      }
      // Lambda subscription — endpoint is the function ARN
      if (sub.protocol === 'lambda' || sub.endpoint?.includes(':lambda:') || sub.endpoint?.includes(':function:')) {
        const fnName = sub.endpoint.split(':').pop();
        // SNS event format: { Records: [{ EventSource: 'aws:sns', Sns: {...} }] }
        const event = { Records: [{ EventSource: 'aws:sns', EventVersion: '1.0', EventSubscriptionArn: sub.subArn, Sns: snsEnvelope }] };
        invokeLambda(fnName, event, { source: 'sns' }).catch(()=>{});
        continue;
      }
      // HTTP/HTTPS/email/SMS — not delivered, but counted as accepted
      // (kept here as a no-op so we don't fail the publish call).
    } catch (e) {
      console.warn(`[SNS] Subscriber delivery failed (${sub.subArn}):`, e.message);
    }
  }
}

// ── Subscription attributes ────────────────────────────────────────────────
function applySubscriptionAttributes(sub, attrs) {
  if (!attrs) return;
  if ('FilterPolicy' in attrs)      sub.filterPolicy = attrs.FilterPolicy || null;
  if ('FilterPolicyScope' in attrs) sub.filterPolicyScope = attrs.FilterPolicyScope || 'MessageAttributes';
  if ('RawMessageDelivery' in attrs) sub.rawMessageDelivery = attrs.RawMessageDelivery === 'true';
}

function findSubscription(subArn) {
  for (const topic of Object.values(store.sns.topics)) {
    const sub = (topic.subscriptions || []).find(s => s.subArn === subArn);
    if (sub) return sub;
  }
  return null;
}

function subMemberXml(s) {
  return `<member><SubscriptionArn>${escapeXml(s.subArn)}</SubscriptionArn><Owner>${s.owner || ACCOUNT}</Owner><Protocol>${escapeXml(s.protocol || '')}</Protocol><Endpoint>${escapeXml(s.endpoint || '')}</Endpoint><TopicArn>${escapeXml(s.topicArn)}</TopicArn></member>`;
}

// ── Filter policy evaluation (subset: exact, anything-but, prefix, numeric,
//    exists) over message attributes (default) or the message body (scope=
//    MessageBody). Returns true when the subscription has no policy. ─────────
function subscriptionMatches(sub, attributes, message) {
  if (!sub.filterPolicy) return true;
  let policy;
  try { policy = JSON.parse(sub.filterPolicy); } catch { return true; }
  if ((sub.filterPolicyScope || 'MessageAttributes') === 'MessageBody') {
    let bodyObj;
    try { bodyObj = JSON.parse(message); } catch { return false; }
    return matchPolicyObject(policy, bodyObj);
  }
  // MessageAttributes scope: look up each key in the attribute map.
  return matchPolicyObject(policy, attributesToPlain(attributes || {}));
}

// Convert internal attribute map to a plain { name: value } object for matching.
function attributesToPlain(attrs) {
  const out = {};
  for (const [name, a] of Object.entries(attrs)) out[name] = attrValueForMatch(a);
  return out;
}

function attrValueForMatch(a) {
  if (!a) return undefined;
  const dt = a.DataType || a.Type || 'String';
  const raw = a.StringValue ?? a.Value ?? a.BinaryValue;
  if (dt.startsWith('Number')) return Number(raw);
  if (dt === 'String.Array') { try { return JSON.parse(raw); } catch { return raw; } }
  return raw;
}

function matchPolicyObject(policy, obj) {
  for (const [key, spec] of Object.entries(policy)) {
    const present = obj != null && Object.prototype.hasOwnProperty.call(obj, key);
    const value = present ? obj[key] : undefined;
    if (Array.isArray(spec)) {
      if (!matchKey(spec, present, value)) return false;
    } else if (spec && typeof spec === 'object') {
      // Nested object → recurse (MessageBody scope only).
      if (!matchPolicyObject(spec, present ? value : undefined)) return false;
    }
  }
  return true;
}

function matchKey(conditions, present, value) {
  const values = Array.isArray(value) ? value : [value];
  return conditions.some(cond => {
    if (cond && typeof cond === 'object' && 'exists' in cond) return cond.exists ? present : !present;
    if (!present) return false;
    return values.some(v => matchCondition(cond, v));
  });
}

function matchCondition(cond, val) {
  if (typeof cond === 'string') return val === cond;
  if (typeof cond === 'number') return Number(val) === cond;
  if (typeof cond === 'boolean') return val === cond;
  if (cond && typeof cond === 'object') {
    if ('anything-but' in cond) {
      const set = Array.isArray(cond['anything-but']) ? cond['anything-but'] : [cond['anything-but']];
      return !set.includes(val);
    }
    if ('prefix' in cond) return typeof val === 'string' && val.startsWith(cond.prefix);
    if ('numeric' in cond) return matchNumeric(cond.numeric, val);
  }
  return false;
}

function matchNumeric(spec, val) {
  const n = Number(val);
  if (Number.isNaN(n)) return false;
  for (let i = 0; i < spec.length; i += 2) {
    const op = spec[i], operand = Number(spec[i + 1]);
    if (op === '=' && !(n === operand)) return false;
    if (op === '<' && !(n < operand)) return false;
    if (op === '<=' && !(n <= operand)) return false;
    if (op === '>' && !(n > operand)) return false;
    if (op === '>=' && !(n >= operand)) return false;
  }
  return true;
}

// ── Form parsers (AWS query protocol flattening) ────────────────────────────
// MessageAttributes.entry.N.Name / .Value.DataType / .Value.StringValue|BinaryValue
function parseMessageAttributes(params, prefix) {
  const out = {};
  for (let i = 1; ; i++) {
    const name = params.get(`${prefix}.entry.${i}.Name`);
    if (name == null) break;
    const dataType = params.get(`${prefix}.entry.${i}.Value.DataType`) || 'String';
    const binary = params.get(`${prefix}.entry.${i}.Value.BinaryValue`);
    out[name] = { DataType: dataType };
    if (binary != null) out[name].BinaryValue = binary;
    else out[name].StringValue = params.get(`${prefix}.entry.${i}.Value.StringValue`) ?? '';
  }
  return Object.keys(out).length ? out : null;
}

// String→String maps: Prefix.entry.N.key / .value
function parseStringMap(params, prefix) {
  const out = {};
  for (let i = 1; ; i++) {
    const k = params.get(`${prefix}.entry.${i}.key`);
    if (k == null) break;
    out[k] = params.get(`${prefix}.entry.${i}.value`) ?? '';
  }
  return Object.keys(out).length ? out : null;
}

// PublishBatchRequestEntries.member.N.{Id,Message,Subject,MessageAttributes...}
function parsePublishBatchEntries(params) {
  const entries = [];
  for (let i = 1; ; i++) {
    const base = `PublishBatchRequestEntries.member.${i}`;
    const id = params.get(`${base}.Id`);
    if (id == null) break;
    entries.push({
      Id: id,
      Message: params.get(`${base}.Message`) || '',
      Subject: params.get(`${base}.Subject`) || '',
      MessageAttributes: parseMessageAttributes(params, `${base}.MessageAttributes`),
    });
  }
  return entries;
}

// ── Misc helpers ────────────────────────────────────────────────────────────
function hasAttributes(attrs) {
  return attrs && typeof attrs === 'object' && Object.keys(attrs).length > 0;
}

// Internal attribute map → SNS envelope form { name: { Type, Value } }
function toEnvelopeAttributes(attrs) {
  const out = {};
  for (const [name, a] of Object.entries(attrs)) {
    out[name] = { Type: a.DataType || 'String', Value: a.BinaryValue ?? a.StringValue ?? '' };
  }
  return out;
}

function attrEntriesXml(map) {
  return Object.entries(map)
    .map(([k, v]) => `<entry><key>${escapeXml(k)}</key><value>${escapeXml(String(v))}</value></entry>`)
    .join('');
}

function wrap(respTag, resultTag, inner) {
  return `<?xml version="1.0"?><${respTag} xmlns="http://sns.amazonaws.com/doc/2010-03-31/"><${resultTag}>${inner}</${resultTag}><ResponseMetadata><RequestId>${randomId(36)}</RequestId></ResponseMetadata></${respTag}>`;
}

File: internal/services/sns/sns.go
// Package sns — port of src/services/sns.js: topics, subscriptions, publish
// fan-out to SQS/Lambda with FilterPolicy evaluation (exact / anything-but /
// prefix / numeric / exists over MessageAttributes or MessageBody scope),
// RawMessageDelivery, PublishBatch. Query protocol only, like Node.
//
// Fan-out is fire-and-forget (Publish returns once accepted, not delivered):
// the handler responds, then a goroutine walks the subscriptions — SQS
// enqueues re-enter the store lock, Lambda invokes never hold it.
package sns

import (
	"encoding/json"
	"net/http"
	"net/url"
	"sort"
	"strconv"
	"strings"

	"github.com/mockcloud/mockcloud/internal/httpapi"
	"github.com/mockcloud/mockcloud/internal/protocol/respond"
	"github.com/mockcloud/mockcloud/internal/services/lambda"
	"github.com/mockcloud/mockcloud/internal/services/sqs"
	"github.com/mockcloud/mockcloud/internal/state"
	"github.com/mockcloud/mockcloud/internal/store"
)

const account = "000000000000"

type Service struct {
	st        *store.Store
	lambdaSvc *lambda.Service
}

func New(st *store.Store, lambdaSvc *lambda.Service) *Service {
	return &Service{st: st, lambdaSvc: lambdaSvc}
}

func wrap(respTag, resultTag, inner string) string {
	return `<?xml version="1.0"?><` + respTag + ` xmlns="http://sns.amazonaws.com/doc/2010-03-31/"><` +
		resultTag + `>` + inner + `</` + resultTag + `><ResponseMetadata><RequestId>` +
		state.RandomID(36) + `</RequestId></ResponseMetadata></` + respTag + `>`
}

func attrEntriesXML(pairs [][2]string) string {
	var sb strings.Builder
	for _, p := range pairs {
		sb.WriteString("<entry><key>" + respond.EscapeXML(p[0]) + "</key><value>" + respond.EscapeXML(p[1]) + "</value></entry>")
	}
	return sb.String()
}

func subMemberXML(s *state.Subscription, topicArn string) string {
	owner := s.Owner
	if owner == "" {
		owner = account
	}
	return "<member><SubscriptionArn>" + respond.EscapeXML(s.SubArn) + "</SubscriptionArn><Owner>" + owner +
		"</Owner><Protocol>" + respond.EscapeXML(s.Protocol) + "</Protocol><Endpoint>" + respond.EscapeXML(s.Endpoint) +
		"</Endpoint><TopicArn>" + respond.EscapeXML(topicArn) + "</TopicArn></member>"
}

func (svc *Service) Handler(w http.ResponseWriter, r *httpapi.Request) {
	params, _ := url.ParseQuery(string(r.RawBody))
	action := r.Query.Get("Action")
	if action == "" {
		action = params.Get("Action")
	}
	get := params.Get

	switch action {
	case "CreateTopic":
		name := get("Name")
		a := state.Arn("sns", name)
		svc.st.With(func(s *state.State) {
			if s.SNS.Topics[a] == nil {
				s.SNS.Topics[a] = &state.Topic{
					Name: name, Arn: a, Created: state.NowMs(),
					Subscriptions: []*state.Subscription{}, Attributes: map[string]string{},
				}
			}
		})
		respond.XML(w, 200, wrap("CreateTopicResponse", "CreateTopicResult", "<TopicArn>"+respond.EscapeXML(a)+"</TopicArn>"))

	case "DeleteTopic":
		svc.st.With(func(s *state.State) { delete(s.SNS.Topics, get("TopicArn")) })
		respond.XML(w, 200, wrap("DeleteTopicResponse", "DeleteTopicResult", ""))

	case "ListTopics":
		var arns []string
		svc.st.With(func(s *state.State) {
			for a := range s.SNS.Topics {
				arns = append(arns, a)
			}
		})
		sort.Strings(arns)
		var sb strings.Builder
		for _, a := range arns {
			sb.WriteString("<member><TopicArn>" + respond.EscapeXML(a) + "</TopicArn></member>")
		}
		respond.XML(w, 200, wrap("ListTopicsResponse", "ListTopicsResult", "<Topics>"+sb.String()+"</Topics>"))

	case "GetTopicAttributes":
		var pairs [][2]string
		var missing bool
		svc.st.With(func(s *state.State) {
			t := s.SNS.Topics[get("TopicArn")]
			if t == nil {
				missing = true
				return
			}
			pairs = [][2]string{
				{"TopicArn", t.Arn}, {"Owner", account},
				{"SubscriptionsConfirmed", strconv.Itoa(len(t.Subscriptions))},
				{"SubscriptionsPending", "0"}, {"SubscriptionsDeleted", "0"},
				{"DisplayName", t.Name},
			}
			names := make([]string, 0, len(t.Attributes))
			for n := range t.Attributes {
				names = append(names, n)
			}
			sort.Strings(names)
			for _, n := range names {
				pairs = append(pairs, [2]string{n, t.Attributes[n]})
			}
		})
		if missing {
			respond.ErrorXML(w, 404, "NotFound", "Topic not found")
			return
		}
		respond.XML(w, 200, wrap("GetTopicAttributesResponse", "GetTopicAttributesResult", "<Attributes>"+attrEntriesXML(pairs)+"</Attributes>"))

	case "SetTopicAttributes":
		var missing bool
		svc.st.With(func(s *state.State) {
			t := s.SNS.Topics[get("TopicArn")]
			if t == nil {
				missing = true
				return
			}
			if name := get("AttributeName"); name != "" {
				t.Attributes[name] = get("AttributeValue")
			}
		})
		if missing {
			respond.ErrorXML(w, 404, "NotFound", "Topic not found")
			return
		}
		respond.XML(w, 200, wrap("SetTopicAttributesResponse", "SetTopicAttributesResult", ""))

	case "Subscribe":
		topicArn := get("TopicArn")
		subArn := topicArn + ":" + state.RandomID(8)
		var missing bool
		svc.st.With(func(s *state.State) {
			t := s.SNS.Topics[topicArn]
			if t == nil {
				missing = true
				return
			}
			sub := &state.Subscription{
				SubArn: subArn, TopicArn: topicArn, Protocol: get("Protocol"),
				Endpoint: get("Endpoint"), Status: "confirmed", Owner: account,
			}
			applySubscriptionAttributes(sub, parseStringMap(params, "Attributes"))
			t.Subscriptions = append(t.Subscriptions, sub)
		})
		if missing {
			respond.ErrorXML(w, 404, "NotFound", "Topic not found")
			return
		}
		respond.XML(w, 200, wrap("SubscribeResponse", "SubscribeResult", "<SubscriptionArn>"+respond.EscapeXML(subArn)+"</SubscriptionArn>"))

	case "SetSubscriptionAttributes":
		var found bool
		svc.st.With(func(s *state.State) {
			sub := findSubscription(s, get("SubscriptionArn"))
			if sub == nil {
				return
			}
			found = true
			if name := get("AttributeName"); name != "" {
				applySubscriptionAttributes(sub, map[string]string{name: get("AttributeValue")})
			}
		})
		if !found {
			respond.ErrorXML(w, 404, "NotFound", "Subscription not found")
			return
		}
		respond.XML(w, 200, wrap("SetSubscriptionAttributesResponse", "SetSubscriptionAttributesResult", ""))

	case "GetSubscriptionAttributes":
		var pairs [][2]string
		svc.st.With(func(s *state.State) {
			sub := findSubscription(s, get("SubscriptionArn"))
			if sub == nil {
				return
			}
			owner := sub.Owner
			if owner == "" {
				owner = account
			}
			pairs = [][2]string{
				{"SubscriptionArn", sub.SubArn}, {"TopicArn", sub.TopicArn},
				{"Protocol", sub.Protocol}, {"Endpoint", sub.Endpoint}, {"Owner", owner},
				{"ConfirmationWasAuthenticated", "true"}, {"PendingConfirmation", "false"},
				{"RawMessageDelivery", boolStr(sub.RawMessageDelivery)},
			}
			if sub.FilterPolicy != nil {
				scope := sub.FilterPolicyScope
				if scope == "" {
					scope = "MessageAttributes"
				}
				pairs = append(pairs, [2]string{"FilterPolicy", *sub.FilterPolicy}, [2]string{"FilterPolicyScope", scope})
			}
		})
		if pairs == nil {
			respond.ErrorXML(w, 404, "NotFound", "Subscription not found")
			return
		}
		respond.XML(w, 200, wrap("GetSubscriptionAttributesResponse", "GetSubscriptionAttributesResult", "<Attributes>"+attrEntriesXML(pairs)+"</Attributes>"))

	case "Publish":
		topicArn := get("TopicArn")
		attributes := parseMessageAttributes(params, "MessageAttributes")
		msgID := state.RandomID(36)
		var missing bool
		svc.st.With(func(s *state.State) {
			t := s.SNS.Topics[topicArn]
			if t == nil {
				missing = true
				return
			}
			t.Published++
		})
		if missing {
			respond.ErrorXML(w, 404, "NotFound", "Topic not found")
			return
		}
		go svc.Fanout(topicArn, msgID, get("Message"), get("Subject"), attributes)
		respond.XML(w, 200, wrap("PublishResponse", "PublishResult", "<MessageId>"+msgID+"</MessageId>"))

	case "PublishBatch":
		topicArn := get("TopicArn")
		var missing bool
		svc.st.With(func(s *state.State) { missing = s.SNS.Topics[topicArn] == nil })
		if missing {
			respond.ErrorXML(w, 404, "NotFound", "Topic not found")
			return
		}
		var successXML strings.Builder
		for _, e := range parsePublishBatchEntries(params) {
			msgID := state.RandomID(36)
			svc.st.With(func(s *state.State) {
				if t := s.SNS.Topics[topicArn]; t != nil {
					t.Published++
				}
			})
			e := e
			go svc.Fanout(topicArn, msgID, e.message, e.subject, e.attributes)
			successXML.WriteString("<member><Id>" + respond.EscapeXML(e.id) + "</Id><MessageId>" + msgID + "</MessageId></member>")
		}
		respond.XML(w, 200, wrap("PublishBatchResponse", "PublishBatchResult",
			"<Successful>"+successXML.String()+"</Successful><Failed/>"))

	case "Unsubscribe":
		subArn := get("SubscriptionArn")
		svc.st.With(func(s *state.State) {
			for _, t := range s.SNS.Topics {
				var kept []*state.Subscription
				for _, sub := range t.Subscriptions {
					if sub.SubArn != subArn {
						kept = append(kept, sub)
					}
				}
				if kept == nil {
					kept = []*state.Subscription{}
				}
				t.Subscriptions = kept
			}
		})
		respond.XML(w, 200, wrap("UnsubscribeResponse", "UnsubscribeResult", ""))

	case "ListSubscriptions", "ListSubscriptionsByTopic":
		byTopic := action == "ListSubscriptionsByTopic"
		var sb strings.Builder
		var missing bool
		svc.st.With(func(s *state.State) {
			if byTopic {
				t := s.SNS.Topics[get("TopicArn")]
				if t == nil {
					missing = true
					return
				}
				for _, sub := range t.Subscriptions {
					sb.WriteString(subMemberXML(sub, t.Arn))
				}
				return
			}
			arns := make([]string, 0, len(s.SNS.Topics))
			for a := range s.SNS.Topics {
				arns = append(arns, a)
			}
			sort.Strings(arns)
			for _, a := range arns {
				t := s.SNS.Topics[a]
				for _, sub := range t.Subscriptions {
					sb.WriteString(subMemberXML(sub, t.Arn))
				}
			}
		})
		if missing {
			respond.ErrorXML(w, 404, "NotFound", "Topic not found")
			return
		}
		tag := "ListSubscriptions"
		if byTopic {
			tag = "ListSubscriptionsByTopic"
		}
		respond.XML(w, 200, wrap(tag+"Response", tag+"Result", "<Subscriptions>"+sb.String()+"</Subscriptions>"))

	default:
		respond.ErrorXML(w, 400, "InvalidAction", "Unknown action: "+action)
	}
}

func boolStr(b bool) string {
	if b {
		return "true"
	}
	return "false"
}

func applySubscriptionAttributes(sub *state.Subscription, attrs map[string]string) {
	if attrs == nil {
		return
	}
	if v, ok := attrs["FilterPolicy"]; ok {
		if v == "" {
			sub.FilterPolicy = nil
		} else {
			sub.FilterPolicy = &v
		}
	}
	if v, ok := attrs["FilterPolicyScope"]; ok {
		if v == "" {
			v = "MessageAttributes"
		}
		sub.FilterPolicyScope = v
	}
	if v, ok := attrs["RawMessageDelivery"]; ok {
		sub.RawMessageDelivery = v == "true"
	}
}

func findSubscription(s *state.State, subArn string) *state.Subscription {
	for _, t := range s.SNS.Topics {
		for _, sub := range t.Subscriptions {
			if sub.SubArn == subArn {
				return sub
			}
		}
	}
	return nil
}

// ── Fan-out (fanoutSnsMessage) ───────────────────────────────────────────────

// Fanout delivers one published message to every matching subscriber.
// Exported for EventBridge SNS targets and S3 notifications.
func (svc *Service) Fanout(topicArn, msgID, message, subject string, attributes map[string]any) {
	type delivery struct {
		sub      state.Subscription
		envelope string
	}
	var deliveries []delivery
	svc.st.With(func(s *state.State) {
		t := s.SNS.Topics[topicArn]
		if t == nil {
			return
		}
		envelope := map[string]any{
			"Type": "Notification", "MessageId": msgID, "TopicArn": t.Arn,
			"Subject": subject, "Message": message, "Timestamp": state.ISO(state.NowMs()),
		}
		if len(attributes) > 0 {
			envelope["MessageAttributes"] = toEnvelopeAttributes(attributes)
		}
		envJSON := string(respond.Marshal(envelope))
		for _, sub := range t.Subscriptions {
			if !subscriptionMatches(sub, attributes, message) {
				continue
			}
			deliveries = append(deliveries, delivery{sub: *sub, envelope: envJSON})
		}
	})

	for _, d := range deliveries {
		sub := d.sub
		switch {
		// SQS subscription — endpoint is the queue ARN.
		case sub.Protocol == "sqs" || strings.Contains(sub.Endpoint, ":sqs:"):
			qurl := sqs.QueueURLForArn(sub.Endpoint)
			svc.st.With(func(s *state.State) {
				if s.SQS.Queues[qurl] == nil {
					return
				}
				if sub.RawMessageDelivery {
					sqs.EnqueueWithAttributesLocked(s, qurl, message, attributes)
				} else {
					sqs.EnqueueJSONLocked(s, qurl, d.envelope)
				}
			})
		// Lambda subscription — endpoint is the function ARN.
		case sub.Protocol == "lambda" || strings.Contains(sub.Endpoint, ":lambda:") || strings.Contains(sub.Endpoint, ":function:"):
			parts := strings.Split(sub.Endpoint, ":")
			fnName := parts[len(parts)-1]
			var env map[string]any
			_ = json.Unmarshal([]byte(d.envelope), &env)
			event := map[string]any{"Records": []any{map[string]any{
				"EventSource": "aws:sns", "EventVersion": "1.0",
				"EventSubscriptionArn": sub.SubArn, "Sns": env,
			}}}
			go svc.lambdaSvc.Invoke(fnName, string(respond.Marshal(event)), "sns", "")
			// HTTP/HTTPS/email/SMS — accepted but not delivered (no-op).
		}
	}
}

// Internal attribute map → SNS envelope form { name: { Type, Value } }.
func toEnvelopeAttributes(attrs map[string]any) map[string]any {
	out := map[string]any{}
	for name, raw := range attrs {
		a, _ := raw.(map[string]any)
		typ, _ := a["DataType"].(string)
		if typ == "" {
			typ = "String"
		}
		value := a["BinaryValue"]
		if value == nil {
			value = a["StringValue"]
		}
		if value == nil {
			value = ""
		}
		out[name] = map[string]any{"Type": typ, "Value": value}
	}
	return out
}

// ── Query-protocol form parsers ──────────────────────────────────────────────

func parseMessageAttributes(params url.Values, prefix string) map[string]any {
	out := map[string]any{}
	for i := 1; ; i++ {
		base := prefix + ".entry." + strconv.Itoa(i)
		if _, present := params[base+".Name"]; !present {
			break
		}
		name := params.Get(base + ".Name")
		dataType := params.Get(base + ".Value.DataType")
		if dataType == "" {
			dataType = "String"
		}
		attr := map[string]any{"DataType": dataType}
		if _, hasBin := params[base+".Value.BinaryValue"]; hasBin {
			attr["BinaryValue"] = params.Get(base + ".Value.BinaryValue")
		} else {
			attr["StringValue"] = params.Get(base + ".Value.StringValue")
		}
		out[name] = attr
	}
	if len(out) == 0 {
		return nil
	}
	return out
}

func parseStringMap(params url.Values, prefix string) map[string]string {
	out := map[string]string{}
	for i := 1; ; i++ {
		base := prefix + ".entry." + strconv.Itoa(i)
		if _, present := params[base+".key"]; !present {
			break
		}
		out[params.Get(base+".key")] = params.Get(base + ".value")
	}
	if len(out) == 0 {
		return nil
	}
	return out
}

type batchEntry struct {
	id, message, subject string
	attributes           map[string]any
}

func parsePublishBatchEntries(params url.Values) []batchEntry {
	var entries []batchEntry
	for i := 1; ; i++ {
		base := "PublishBatchRequestEntries.member." + strconv.Itoa(i)
		if _, present := params[base+".Id"]; !present {
			break
		}
		entries = append(entries, batchEntry{
			id: params.Get(base + ".Id"), message: params.Get(base + ".Message"),
			subject:    params.Get(base + ".Subject"),
			attributes: parseMessageAttributes(params, base+".MessageAttributes"),
		})
	}
	return entries
}
